File: solutions/sum-root-to-leaf-numbers/sum-root-to-leaf-numbers.c
// @Title: Sum Root to Leaf Numbers
// @Author: colinjxc
// @Date: 2022-02-06T22:19:47+08:00
// @URL: https://leetcode-cn.com/problems/sum-root-to-leaf-numbers


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
void sumNumbersDfs(struct TreeNode* root, int sumPrev, int *total) {
    if (root == NULL) return;
    int num = sumPrev*10 + root->val;
    if (root->left == NULL && root->right == NULL) {
        *total += num;
        return;
    }
    if (root->left) sumNumbersDfs(root->left, num, total);
    if (root->right) sumNumbersDfs(root->right, num, total);
}

int sumNumbers(struct TreeNode* root){
    int sum = 0;
    sumNumbersDfs(root, 0, &sum);
    return sum;
}

File: solutions/sum-root-to-leaf-numbers/sum-root-to-leaf-numbers.go
// @Title: Sum Root to Leaf Numbers
// @Author: colinjxc
// @Date: 2022-02-06T22:17:23+08:00
// @URL: https://leetcode-cn.com/problems/sum-root-to-leaf-numbers


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */


func sumNumbers(root *TreeNode) int {
    return helper(root, 0)
}

func helper(root *TreeNode, prev int) int {
    if root == nil {
        return 0
    }
    sum := 10*prev + root.Val
    if root.Left == nil && root.Right == nil {
        return sum
    }
    return helper(root.Left, sum) + helper(root.Right, sum)
}
